tls/tls_ticket.c

```diff
@@ -22,409 +22,646 @@
+	int r;
 
-	if (ctx == NULL)
-		return(TTLS_ERR_BAD_INPUT_DATA);
+	memcpy_fast(tik->name, key->name, sizeof(key->name));
+	ttls_rnd(tik->iv, sizeof(tik->iv));
 
-	/* See ttls_ticket_write() */
-	if (len < 4 + 12 + 2 + 16)
-		return(TTLS_ERR_BAD_INPUT_DATA);
+	if ((r = ttls_ticket_ctx_init(&ctx, tik, key)))
+		return r;
 
-	spin_lock(&ctx->mutex);
+	aead_request_set_crypt(ctx.req, sgt.sgl, sgt.sgl, crypt_len, tik->iv);
 
-	if ((ret = ssl_ticket_update_keys(ctx)) != 0)
-		goto cleanup;
+	sg_init_table(&sg, 1);
+	sg_set_buf(&sg, tik, crypt_len);
+	if ((r = crypto_aead_encrypt(ctx.req)))
+		T_WARN("AEAD encryption failed: %d\n", r);
 
-	enc_len = (enc_len_p[0] << 8) | enc_len_p[1];
-	tag = ticket + enc_len;
+	ttls_ticket_ctx_free(&ctx);
 
-	if (len != 4 + 12 + 2 + enc_len + 16)
-	{
-		ret = TTLS_ERR_BAD_INPUT_DATA;
-		goto cleanup;
-	}
+	return r;
+}
 
-	/* Select key */
-	if ((key = ssl_ticket_select_key(ctx, key_name)) == NULL)
-	{
-		/* We can't know for sure but this is a likely option unless we're
-		 * under attack - this is only informative anyway */
-		ret = TTLS_ERR_SESSION_TICKET_EXPIRED;
-		goto cleanup;
-	}
+static int
+ttls_ticket_decrypt(TlsTicket *tik, size_t len, TlsTicketKey *key)
+{
+	TlSTicketCryptCtx ctx = { 0 };
+	size_t decrypt_len;
+	struct scatterlist sg;
+	struct sg_table sgt = {
+		.sgl	= &sg,
+		.nents	= 1,
+	};
+	int r;
 
-	/* Decrypt and authenticate */
-	/* TODO #1054 replace with linux/crypto as in ttls.c. */
-	if ((ret = ttls_cipher_auth_decrypt(&key->ctx, iv, 12,
-		key_name, 4 + 12 + 2, ticket, enc_len,
-		ticket, &clear_len, tag, 16)) != 0)
-	{
-		goto cleanup;
-	}
-	if (clear_len != enc_len)
-	{
-		ret = TTLS_ERR_INTERNAL_ERROR;
-		goto cleanup;
-	}
+	if ((r = ttls_ticket_ctx_init(&ctx, tik, key)))
+		return r;
 
-	/* Actually load session */
-	if ((ret = ssl_load_session(session, ticket, clear_len)) != 0)
-		goto cleanup;
+	decrypt_len = len - (sizeof(TlsTicket) - sizeof(TlsState));
+	aead_request_set_crypt(ctx.req, sgt.sgl, sgt.sgl, decrypt_len, tik->iv);
 
-	{
-		/* Check for expiration */
-		time_t current_time = ttls_time();
-
-		if (current_time < session->start ||
-			(uint32_t)(current_time - session->start) > ctx->ticket_lifetime)
-		{
-			ret = TTLS_ERR_SESSION_TICKET_EXPIRED;
-			goto cleanup;
-		}
-	}
+	sg_init_table(&sg, 1);
+	sg_set_buf(&sg, tik, decrypt_len);
+	if ((r = crypto_aead_decrypt(ctx.req)))
+		T_WARN("AEAD decryption failed: %d\n", r);
 
-cleanup:
-	spin_unlock(&ctx->mutex);
+	ttls_ticket_ctx_free(&ctx);
 
-	return ret;
+	return r;
 }
 
-/*
- * Free context
+/**
+ * Parse a session ticket as generated by the @ttls_ticket_write() function,
+ * and, if the ticket is authentic and valid, load the session.
  */
-void ttls_ticket_free(ttls_ticket_context *ctx)
+int
+ttls_ticket_parse(TlsCtx *ctx, unsigned char *buf, size_t len)
 {
-	ttls_cipher_free(&ctx->keys[0].ctx);
-	ttls_cipher_free(&ctx->keys[1].ctx);
-	bzero_fast(ctx, sizeof(ttls_ticket_context));
+	TlsTicket *tik = (TlsTicket *)buf;
```

Comment:
It _is_ scary, I don't see any reason to hide `char *` over the `void *`. This data comes from network (from a user) and we really match raw data with C-structure here. Before the call we assured that the raw data has sufficient size and not split by segments, in the same time the structure has a `packed` attribute. 

Since the data in not used by other agents except Tempesta itself, we don't need to care about cross platform compatibility and network-to-host data transformation and can reuse the buffer as is. If we can decrypt it of course.

Reply:
Technically this might yield undefined behavior, I believe. There might be lots of places like this in the code. Perhaps, it makes sense to try and compile the whole codebase with **-fstrict-aliasing -Wstrict-aliasing -std=c99**? Just assess the scope of the issue.

However, the kernel is intentionally, I believe, compiled with **-fno-strict-aliasing**. See this, well, expressive one might say, speech by Linus: [Linus on strict aliasing](https://www.yodaiken.com/2018/06/07/torvalds-on-aliasing/). Linus has changed since then and promised to tone down his language. Anyway, in his little speech he does say how type punning should be done right.